core/src/impl/Kokkos_SharedAlloc_timpl.hpp

```diff
@@ -0,0 +1,285 @@
+      allocate(r_old->m_space, r_old->get_label(), arg_alloc_size);
+
+  Kokkos::Impl::DeepCopy<MemorySpace, MemorySpace>(
+      r_new->data(), r_old->data(), std::min(r_old->size(), r_new->size()));
+
+  record_base_t::increment(r_new);
+  record_base_t::decrement(r_old);
+
+  return r_new->data();
+}
+
+template <class MemorySpace>
+auto SharedAllocationRecordCommon<MemorySpace>::get_record(void* alloc_ptr)
+    -> derived_t* {
+  using Header = SharedAllocationHeader;
+
+  Header const* const h = alloc_ptr ? Header::get_header(alloc_ptr) : nullptr;
+
+  if (!alloc_ptr || h->m_record->m_alloc_ptr != h) {
+    using namespace std::literals::string_literals;
+    Kokkos::Impl::throw_runtime_exception(
+        "Kokkos::Impl::SharedAllocationRecordCommon<"s +
+        std::string(MemorySpace::name()) + ">::get_record() ERROR"s);
+  }
+
+  return static_cast<derived_t*>(h->m_record);
+}
+
+template <class MemorySpace>
+std::string SharedAllocationRecordCommon<MemorySpace>::get_label() const {
+  return std::string(record_base_t::head()->m_label);
+}
+
+template <class MemorySpace>
+void SharedAllocationRecordCommon<MemorySpace>::
+    _fill_host_accessible_header_info(SharedAllocationHeader& arg_header,
+                                      std::string const& arg_label) {
+  // Fill in the Header information, directly accessible on the host
+
+  arg_header.m_record = &self();
+
+  strncpy(arg_header.m_label, arg_label.c_str(),
+          SharedAllocationHeader::maximum_label_length);
+  // Set last element zero, in case c_str is too long
+  arg_header.m_label[SharedAllocationHeader::maximum_label_length - 1] = '\0';
+}
+
+template <class MemorySpace>
+void SharedAllocationRecordCommon<MemorySpace>::print_records(
+    std::ostream& s, const MemorySpace&, bool detail) {
+  (void)s;
+  (void)detail;
+#ifdef KOKKOS_ENABLE_DEBUG
+  SharedAllocationRecord<void, void>::print_host_accessible_records(
+      s, MemorySpace::name(), &derived_t::s_root_record, detail);
+#else
+  using namespace std::literals::string_literals;
+  Kokkos::Impl::throw_runtime_exception(
+      "SharedAllocationHeader<"s + std::string(MemorySpace::name()) +
+      ">::print_records only works with KOKKOS_ENABLE_DEBUG enabled"s);
+#endif
+}
+
+template <class MemorySpace>
+void HostInaccessibleSharedAllocationRecordCommon<MemorySpace>::print_records(
+    std::ostream& s, const MemorySpace&, bool detail) {
+  (void)s;
+  (void)detail;
+#ifdef KOKKOS_ENABLE_DEBUG
+  SharedAllocationRecord<void, void>* r = &derived_t::s_root_record;
+
+  char buffer[256];
+
+  SharedAllocationHeader head;
+
+  if (detail) {
+    do {
+      if (r->m_alloc_ptr) {
+        Kokkos::Impl::DeepCopy<HostSpace, MemorySpace>(
+            &head, r->m_alloc_ptr, sizeof(SharedAllocationHeader));
+      } else {
+        head.m_label[0] = 0;
+      }
+
+      // Formatting dependent on sizeof(uintptr_t)
+      const char* format_string;
+
+      if (sizeof(uintptr_t) == sizeof(unsigned long)) {
```

Comment:
Why can't you just use `%p` https://godbolt.org/z/vszj3o

Reply:
Better https://godbolt.org/z/PfG3Eo, also `%#x` means you don't have to write `0x` in front of of the input.

Reply:
Probably, but this is old code and a different issue from the current pull request. I just moved it over from the existing code that was duplicated across several backends. Either way, if we want to change it, we shouldn't do it here.

Reply:
I'm skeptical of that argument.  You just touched the lines and moved them as part of a clean up.  I'm not saying you have to fix it, but now blame will imply that you wrote it this way instead of whoever wrote the code you copied and pasted.